File: request-reply/src/broker/cliente.java
import org.zeromq.ZMQ;
import org.zeromq.ZContext;

class cliente { 
    public static void main(String[] args) {
        try (ZContext context = new ZContext()) {
            //ZeroMQ
            ZMQ.Socket socket = context.createSocket(ZMQ.REQ);
            socket.connect("tcp://broker:5555");
            
            System.out.println(" Cliente Java Iniciado");
            
            // Bot automático
            boolean loggedIn = false;
            String username = "bot_java_1";
            
            // Retry em caso de erro
            while (!loggedIn) {
                try {
                    Message loginMsg = new Message("login");
                    loginMsg.setUsername(username);
                    
                    byte[] msgBytes = MessagePackUtil.serialize(loginMsg);
                    
                    System.out.println("\nENVIANDO LOGIN!!");
                    System.out.println("Tipo: " + loginMsg.getType());
                    System.out.println("Username: " + loginMsg.getUsername());
                    System.out.println("Timestamp: " + loginMsg.getTimestamp());
                    System.out.println("Bytes enviados: " + msgBytes.length);
                    
                    // Enviar pelo ZeroMQ
                    socket.send(msgBytes, 0);
                    
                    // Aguardar UMA resposta
                    byte[] responseBytes = socket.recv(0);
                    Response response = MessagePackUtil.deserialize(responseBytes, Response.class);
                  
                    System.out.println("\nRESPOSTA RECEBIDA!!");
                    System.out.println("Success: " + response.isSuccess());
                    System.out.println("Message: " + response.getMessage());
                    System.out.println("Timestamp: " + response.getTimestamp());
                    
                    if (response.isSuccess()) {
                        loggedIn = true;
                        System.out.println("\nLOGIN BEM-SUCEDIDO UHULLL\n");
                    } else {
                        System.out.println("\n ERRO NO LOGIN - Tentando novamente em 2s\n");
                        Thread.sleep(2000);
                    }
                    
                } catch (Exception e) {
                    System.err.println("Erro: " + e.getMessage());
                    Thread.sleep(2000);
                }
            }
            
            System.out.println("Bot Logado com Sucesso");
    
            System.out.println("CRIANDO CANAL");
            
            String nomeCanal = "canal-java-bot"; 
            try {
                // mensagem de criação de canal
                Message createChannelMsg = new Message("create_channel");
                createChannelMsg.setUsername(username);  // Enviar username para validação
                createChannelMsg.setChannelName(nomeCanal);
                
                byte[] msgBytes = MessagePackUtil.serialize(createChannelMsg);
                
                System.out.println("\nENVIAR CRIAÇÃO DE CANAL");
                System.out.println("Tipo: " + createChannelMsg.getType());
                System.out.println("Username: " + createChannelMsg.getUsername());
                System.out.println("Nome do Canal: " + createChannelMsg.getChannelName());
                System.out.println("Timestamp: " + createChannelMsg.getTimestamp());
                System.out.println("Bytes enviados: " + msgBytes.length);
                
                socket.send(msgBytes, 0);
                
                byte[] responseBytes = socket.recv(0);
                Response response = MessagePackUtil.deserialize(responseBytes, Response.class);
                
                System.out.println("\nRESPOSTA RECEBIDA");
                System.out.println("Success: " + response.isSuccess());
                System.out.println("Message: " + response.getMessage());
                System.out.println("Channel Name: " + response.getChannelName());
                System.out.println("Timestamp: " + response.getTimestamp());
                
                if (response.isSuccess()) {
                    System.out.println("\nCANAL CRIADO COM SUCESSO: " + response.getChannelName());
                } else {
                    System.out.println("\nERRO AO CRIAR CANAL: " + response.getMessage());
                }
                
            } catch (Exception e) {
                System.err.println("Erro ao criar canal: " + e.getMessage());
                e.printStackTrace();
            }
            
            System.out.println("\nCliente Finalizado :D");
            
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: request-reply/src/broker/cliente.py
import zmq
from time import sleep
import json #eniar json do cliente p servidor
from datetime import datetime
import msgpack
import zoneinfo
from dataclasses import dataclass, asdict
@dataclass
class Message:
    type: str
    username: str
    timestamp: int = 0
    
    def __post_init__(self):
        if self.timestamp == 0:
            self.timestamp = int(datetime.now().timestamp() * 1000)
    
    def pack(self) -> bytes:
        return msgpack.packb(asdict(self))

@dataclass
class Response:
    success: bool
    message: str
    timestamp: int
    
    @staticmethod
    def unpack(data: bytes) -> "Response":
        d = msgpack.unpackb(data)
        return Response(**d)


context = zmq.Context()
socket = context.socket(zmq.REQ)
socket.connect("tcp://broker:5555")
fuso = zoneinfo.ZoneInfo("America/Sao_Paulo")
i = 0
while True:

    msg= msgpack.packb({"type": "login", "username":"loba", "timestamp": datetime.now(tz=fuso).timestamp()*1000})
    print(f"Mensagem {i}:", end=" ", flush=True)#imprimir no terminal
    socket.send(msg)#envia pro servidor
    #mensagem = socket.recv()#rcebe do servidor
    #print(f"{msgpack.unpackb(mensagem)}") #Print sem ter os bytes da mensagem
    mensagem = Response.unpack(socket.recv())
    print(mensagem.message) #Print sem ter os bytes da mensagem
    sleep(0.5)


    """ 
    msg= msgpack.packb({"type": "criar", "username":"eba", "timestamp": datetime.now(tz=fuso).timestamp()*1000})
    print(f"Mensagem {i}:", end=" ", flush=True)#imprimir no terminal
    socket.send(msg)#envia pro servidor
    mensagem = socket.recv()#rcebe do servidor
    print(f"{msgpack.unpackb(mensagem)}") #Print sem ter os bytes da mensagem
    mensagem = Message.unpack(socket.recv())
    print(mensagem) #Print sem ter os bytes da mensagem
    sleep(0.5)

    #para mandar mais mensagens:
    msg= msgpack.packb({"type": "remover", "username":"eba2", "timestamp": datetime.now(tz=fuso).timestamp()*1000})
    print(f"Mensagem {i}:", end=" ", flush=True)#imprimir no terminal
    socket.send(msg)#envia pro servidor
    mensagem = socket.recv()#rcebe do servidor
    print(f"{msgpack.unpackb(mensagem)}") #Print sem ter os bytes da mensagem
    mensagem = Message.unpack(socket.recv())
    print(mensagem) #Print sem ter os bytes da mensagem
    sleep(0.5)
    
    
    #para mandar mais mensagens:
    msg= msgpack.packb({"type": "criar", "username":"eba3", "timestamp": datetime.now(tz=fuso).timestamp()*1000})
    print(f"Mensagem {i}:", end=" ", flush=True)#imprimir no terminal
    socket.send(msg)#envia pro servidor
    mensagem = socket.recv()#rcebe do servidor
    print(f"{msgpack.unpackb(mensagem)}") #Print sem ter os bytes da mensagem
    mensagem = Message.unpack(socket.recv())
    print(mensagem) #Print sem ter os bytes da mensagem
    sleep(0.5)
    

    #para mandar mais mensagens:
    msg= msgpack.packb({"type": "listar", "username":"eba4", "timestamp": datetime.now(tz=fuso).timestamp()*1000})
    print(f"Mensagem {i}:", end=" ", flush=True)#imprimir no terminal
    socket.send(msg)#envia pro servidor
    mensagem = socket.recv()#rcebe do servidor
    print(f"{msgpack.unpackb(mensagem)}") #Print sem ter os bytes da mensagem
    mensagem = Message.unpack(socket.recv())
    print(mensagem) #Print sem ter os bytes da mensagem
    sleep(0.5)
    """
    i += 1
